README.md
###RexRay Bosh Release

REX-Ray delivers persistent storage access for container runtimes, such as Docker and Mesos, and provides an easy interface for enabling advanced storage functionality across common storage, virtualization and cloud platforms. 

This Bosh release only contains the RexRay Release Binary version 0.4, which is required to be placed in the client VM in order to access volumes hosted on any storage platform.

###Creating Bosh Release
 
 1. Obtain code from Github: 
 2. Go into the directory: 
 3. Create Bosh Release
 
 ```
 git clone https://github.com/EMC-CMD/rexray-boshrelease.git
 cd rexray-boshrelease
 bosh create release
 ```
###Uploading Bosh Release
 1. Upload Bosh Release
 
 ``` bosh upload release ```
 
###Configuring RexRay with a BOSH Manifest
In order to run RexRay, it requires a Configuration YMl. Since we are deploying RexRay through BOSH, this YML must go into the deployment manifest underneath the properties section. Below is an example of using this release with ScaleIO (This required the ScaleIO SDC Release as well)

```
rexray:
  modules:
    default-admin:
      desc: The default admin module.
      disabled: false
      host: tcp://127.0.0.1:7979
      type: admin
    default-docker:
      desc: The default docker module.
      disabled: false
      host: tcp://127.0.0.1:9000
      spec: /var/vcap/data/voldrivers/rexray.spec 
      type: docker
  storageDrivers:
  - scaleio
scaleio:
  endpoint: https://*.*.*.*/api
  insecure: false
  userName: admin
  password: password
  protectionDomainID: 4c9e0aa100000000
  protectionDomainName: default    
  storagePoolName: default    
  storagepoolID: 1234567800000000
  systemID: 5c4c45f02c98552e
  systemName: system
  thinOrThick: ThinProvisioned
  useCerts: true
linux:
  volume:
    fileMode: 0777  
```

For more examples of information about using other storage providers, please see <http://rexray.readthedocs.org/>

###Using RexRay for EMC Persistence with CloudFoundry (Diego Release)
In order to use EMC storage with CloudFoundry you will need to deploy Diego cells (also with BOSH) where your applications will then live. 

When deploying Deigo cells the volume manager needs to understand where the volume driver (RexRay) lives. You have two options for configuring this.

-  Change the `default-docker.spec` line in the rexray config above to `/var/vcap/data/voldrivers/rexray.spec`
-  Add `/etc/docker/plugins/rexray.spec` or your specific `*.spec` file for your volume driver to `properties.diego.executor.volman.driver_paths` in your Diego manifest.  
  
###Contact
- For questions on using the Bosh release:
	- Email: [EMCdojo@emc.com](mailto:EMCdojo@emc.com) 
	- Twitter: [@EMCDojo](https://twitter.com/hashtag/emcdojo)
	- Blog: [EMC Dojo Blog](dojoblog.emc.com)
	- Slack Channel:
  		- Organization: <http://codecommunity.slack.com>
  		- Channel: `#project-rexray`
  
	- Github: <https://github.com/emccode/rexray>

 
 
 
 

 


CHANGELOG.md

## v31
c5e3a9c Change CHANGELOG.md
## v32
37bed90 export GOROOT, and PATH


ci/tasks/vats.rb
#! /usr/bin/env ruby
require 'erb'
require_relative 'utils'

class VATs
  include Utils

  VATS_MANIFEST_ERB_PATH = 'rexray-boshrelease/ci/templates/vats-manifest.yml.erb'
  VATS_MANIFEST_YML_PATH = 'rexray-boshrelease/ci/templates/vats-manifest.yml'
  RUN_VATS_SH_PATH = 'rexray-boshrelease/ci/templates/run_vats.sh'
  REXRAY_CONFIG_ERB_PATH = 'rexray-boshrelease/ci/templates/rexray_config.json.erb'
  REXRAY_CONFIG_JSON_PATH = 'rexray-boshrelease/ci/templates/rexray_config.json'

  ENV_PARAMS = %W{
    BOSH_DIRECTOR_PUBLIC_IP
    BOSH_PASSWORD
    BOSH_USER
    REXRAY_RELEASE_NAME
    STORAGE_SERVICE_TYPE
    VATS_DEPLOYMENT_IP
    VATS_DEPLOYMENT_NAME
    VATS_DEPLOYMENT_PASSWORD
  }

  SCALEIO_PARAMS = %W{
    SCALEIO_ENDPOINT
    SCALEIO_INSECURE
    SCALEIO_MDM_IPS
    SCALEIO_PASSWORD
    SCALEIO_STORAGE_POOL_NAME
    SCALEIO_USERNAME
    SCALEIO_VERSION
  }

  ISILON_PARAMS = %W{
    ISILON_DATA_SUBNET
    ISILON_ENDPOINT
    ISILON_INSECURE
    ISILON_PASSWORD
    ISILON_USERNAME
    ISILON_VOLUME_PATH
  }

  def initialize
    @failed = false
    check_env_params(ENV_PARAMS)
    check_env_params(SCALEIO_PARAMS) if @storage_service_type == 'scaleio'
    check_env_params(ISILON_PARAMS) if @storage_service_type == 'isilon'
  end

  def perform
    exec_cmd('apt-get -y update && apt-get install -y sshpass')
    exec_cmd('gem install bosh_cli --no-ri --no-rdoc')
    exec_cmd("bosh target #{@bosh_director_public_ip}")
    exec_cmd("bosh login #{@bosh_user} #{@bosh_password}")

    cleanup

    output = exec_cmd('bosh status --uuid')
    bosh_director_uuid = output[1].strip
    generate_bosh_manifest(bosh_director_uuid)

    upload_bosh_releases
    exec_cmd("bosh deployment #{VATS_MANIFEST_YML_PATH}")
    exec_cmd('bosh -n deploy')

    generate_raxray_config
    scp(REXRAY_CONFIG_JSON_PATH)
    scp(RUN_VATS_SH_PATH)
    ssh_run_vats
  end

  def cleanup
    exec_cmd("bosh -n delete deployment #{@vats_deployment_name}")
    exec_cmd("bosh -n delete release #{@rexray_release_name} || true")
    exec_cmd("bosh -n delete release scaleio-boshrelease || true")
  end

  def scp(filepath)
    exec_cmd("sshpass -p #{@vats_deployment_password} \
              scp -o StrictHostKeyChecking=no #{filepath} \
              vcap@#{@vats_deployment_ip}:/home/vcap/")
  end

  def ssh_run_vats
    exec_cmd("sshpass -p #{@vats_deployment_password} \
              ssh -o StrictHostKeyChecking=no vcap@#{@vats_deployment_ip} \
              \"echo #{@vats_deployment_password} | sudo -S bash -c '/home/vcap/run_vats.sh'\"")
  end

  def upload_bosh_releases
    exec_cmd("pushd rexray-boshrelease && \
              bosh -n create release --force --name #{@rexray_release_name} && \
              bosh -n upload release && \
              popd")

    if @storage_service_type == 'scaleio'
      exec_cmd("pushd scaleio-boshrelease && \
                bosh -n create release --force --name scaleio-boshrelease && \
                bosh -n upload release && \
                popd")
    end
  end

  def generate_bosh_manifest(bosh_director_uuid)
    results = ERB.new(File.read(VATS_MANIFEST_ERB_PATH)).result(binding())
    File.open(VATS_MANIFEST_YML_PATH, 'w+') do |f|
      f.write(results)
    end
  end

  def generate_raxray_config
    results = ERB.new(File.read(REXRAY_CONFIG_ERB_PATH)).result(binding())
    File.open(REXRAY_CONFIG_JSON_PATH, 'w+') do |f|
      f.write(results)
    end
  end

  if __FILE__ == $0
    VATs.new.perform
  end
end


ci/tasks/vats-isilon.sh
#!/usr/bin/env bash

set -e -x
source rexray-bosh-release/ci/tasks/utils.sh

check_param BOSH_DIRECTOR_PUBLIC_IP
check_param BOSH_PASSWORD
check_param BOSH_USER
check_param DEPLOYMENT_PASSWORD
check_param FAKE_VOLUME_NAME
check_param REXRAY_RELEASE_NAME
check_param VOLUME_DRIVER_ADDRESS

check_param ISILON_VATS_DEPLOYMENT_NAME
check_param ISILON_ENDPOINT
check_param ISILON_INSECURE
check_param ISILON_USERNAME
check_param ISILON_PASSWORD
check_param ISILON_VOLUME_PATH
check_param ISILON_ENDPOINT
check_param ISILON_DATA_SUBNET
check_param ISILON_VSPHERE_VM_IP

bosh target ${BOSH_DIRECTOR_PUBLIC_IP}
bosh login ${BOSH_USER} ${BOSH_PASSWORD}

cat > scaleio-acceptance-manifest.yml <<EOF
---
name: ${ISILON_VATS_DEPLOYMENT_NAME}
director_uuid: cd0eb8bc-831e-447d-99c1-9658c76e7721
stemcells:
- alias: trusty
  os: ubuntu-trusty
  version: '3215'
releases:
- name: rexray-release
  version: latest
jobs:
- name: ${ISILON_VATS_DEPLOYMENT_NAME}
  instances: 1
  templates:
  - name: rexray_service
    release: rexray-release
  vm_type: medium
  stemcell: trusty
  azs:
  - z1
  networks:
  - name: private
    static_ips:
    - ${ISILON_VSPHERE_VM_IP}
  properties:
    network_name: private
properties:
  rexray: |
    ---
    rexray:
      modules:
        isilon:
          disabled: false
          host: tcp://127.0.0.1:9000
          spec: /var/vcap/data/voldrivers/rexray_isilon.spec
          http:
            writetimeout: 900
            readtimeout: 900
          type: docker
          libstorage:
            service: isilon
      libstorage:
        embedded: true
        server:
          services:
            isilon:
              driver: isilon
    isilon:
      endpoint: https://${ISILON_ENDPOINT}:8080
      insecure: ${ISILON_INSECURE}
      username: ${ISILON_USERNAME}
      password: ${ISILON_PASSWORD}
      volumePath: ${ISILON_VOLUME_PATH}
      nfsHost: ${ISILON_ENDPOINT}
      dataSubnet: ${ISILON_DATA_SUBNET}
      quotas: false
      sharedMounts: true
    linux:
      volume:
        fileMode: 0777
update:
  canaries: 1
  max_in_flight: 3
  canary_watch_time: 30000-600000
  update_watch_time: 5000-600000

EOF
cat scaleio-acceptance-manifest.yml
bosh deployment scaleio-acceptance-manifest.yml

apt-get -y update && apt-get -y install sshpass

pushd rexray-bosh-release
  bosh -n create release --force --name ${REXRAY_RELEASE_NAME}
  bosh -n upload release
popd

function cleanUp {
  local status=$?
  bosh -n delete deployment ${ISILON_VATS_DEPLOYMENT_NAME}
  bosh -n delete release ${REXRAY_RELEASE_NAME} || true
  exit $status
}
trap cleanUp EXIT

bosh -n deploy

function ssh_run() {
  sshpass -p ${DEPLOYMENT_PASSWORD} ssh -o "StrictHostKeyChecking no" vcap@${ISILON_VSPHERE_VM_IP} \
    "echo ${DEPLOYMENT_PASSWORD} | sudo -S bash -c ' $* '"
}

cat > config.json <<EOF
{
  "volman_driver_path": "/etc/docker/plugins",
  "driver_address": "${VOLUME_DRIVER_ADDRESS}",
  "driver_name": "rexray",
  "create_config": {
    "Name": "${FAKE_VOLUME_NAME}",
    "Opts": {}
  }
}
EOF

sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no config.json vcap@${ISILON_VSPHERE_VM_IP}:/home/vcap/

cat > run_test.sh <<EOF
set -x

add-apt-repository -y ppa:ubuntu-lxc/lxd-stable
apt-get -y update
apt-get -y install golang
apt-get -y install git

mkdir -p gocode
export GOPATH=/home/vcap/gocode
export PATH=\$PATH:\$GOPATH/bin

go get --insecure -f -u gopkg.in/yaml.v2
go get --insecure -f -u github.com/onsi/ginkgo/ginkgo
go get --insecure -f -u github.com/onsi/gomega
go get --insecure -f -u github.com/cloudfoundry-incubator/volume_driver_cert
go get --insecure -f -u code.cloudfoundry.org/clock
go get --insecure -f -u code.cloudfoundry.org/cfhttp
go get --insecure -f -u code.cloudfoundry.org/cfhttp/handlers
go get --insecure -f -u github.com/cloudfoundry/gunk/http_wrap

cd \$GOPATH/src/github.com/cloudfoundry-incubator/volume_driver_cert

export FIXTURE_FILENAME=/home/vcap/config.json
ginkgo -r
EOF

chmod +x run_test.sh
sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no run_test.sh vcap@${ISILON_VSPHERE_VM_IP}:/home/vcap/

ssh_run ./run_test.sh


ci/tasks/vats-scaleio.sh
#!/usr/bin/env bash

set -e -x
source rexray-bosh-release/ci/tasks/utils.sh

check_param BOSH_DIRECTOR_PUBLIC_IP
check_param BOSH_PASSWORD
check_param BOSH_USER
check_param DEPLOYMENT_PASSWORD
check_param FAKE_VOLUME_NAME
check_param REXRAY_RELEASE_NAME
check_param VOLUME_DRIVER_ADDRESS

check_param SCALEIO_ENDPOINT
check_param SCALEIO_INSECURE
check_param SCALEIO_MDM_IPS
check_param SCALEIO_PASSWORD
check_param SCALEIO_PROTECTION_DOMAIN_ID
check_param SCALEIO_PROTECTION_DOMAIN_NAME
check_param SCALEIO_SDC_RELEASE_NAME
check_param SCALEIO_STORAGE_POOL_ID
check_param SCALEIO_STORAGE_POOL_NAME
check_param SCALEIO_SYSTEM_ID
check_param SCALEIO_USER_ID
check_param SCALEIO_USERNAME
check_param SCALEIO_VATS_DEPLOYMENT_NAME
check_param SCALEIO_VERSION
check_param SCALEIO_VSPHERE_VM_IP

bosh target ${BOSH_DIRECTOR_PUBLIC_IP}
bosh login ${BOSH_USER} ${BOSH_PASSWORD}

cat > scaleio-acceptance-manifest.yml <<EOF
---
name: ${SCALEIO_VATS_DEPLOYMENT_NAME}
director_uuid: cd0eb8bc-831e-447d-99c1-9658c76e7721
stemcells:
- alias: trusty
  os: ubuntu-trusty
  version: '3215'
releases:
- name: rexray-release
  version: latest
- name: scaleio-sdc-bosh-release
  version: latest
jobs:
- name: ${SCALEIO_VATS_DEPLOYMENT_NAME}
  instances: 1
  templates:
  - name: setup_sdc
    release: scaleio-sdc-bosh-release
  - name: rexray_service
    release: rexray-release
  vm_type: medium
  stemcell: trusty
  azs:
  - z1
  networks:
  - name: private
    static_ips:
    - ${SCALEIO_VSPHERE_VM_IP}
  properties:
    network_name: private
properties:
  scaleio:
    mdm:
      ips: ${SCALEIO_MDM_IPS}
  rexray: |
    ---
    rexray:
      modules:
        scaleio:
          disabled: false
          host: tcp://127.0.0.1:9000
          spec: /var/vcap/data/voldrivers/rexray_scaleio.spec
          http:
            writetimeout: 900
            readtimeout: 900
          type: docker
          libstorage:
            service: scaleio
      libstorage:
        embedded: true
        server:
          services:
            scaleio:
              driver: scaleio
    scaleio:
      endpoint: ${SCALEIO_ENDPOINT}
      insecure: ${SCALEIO_INSECURE}
      password: ${SCALEIO_PASSWORD}
      protectionDomainName: default
      storagePoolName: ${SCALEIO_STORAGE_POOL_NAME}
      userName: ${SCALEIO_USERNAME}
      version: ${SCALEIO_VERSION}
    linux:
      volume:
        fileMode: 0777
update:
  canaries: 1
  max_in_flight: 3
  canary_watch_time: 30000-600000
  update_watch_time: 5000-600000

EOF
cat scaleio-acceptance-manifest.yml
bosh deployment scaleio-acceptance-manifest.yml

apt-get -y update && apt-get -y install sshpass

pushd scaleio-sdc-bosh-release
  bosh -n create release --force --name ${SCALEIO_SDC_RELEASE_NAME}
  bosh -n upload release
popd

pushd rexray-bosh-release
  bosh -n create release --force --name ${REXRAY_RELEASE_NAME}
  bosh -n upload release
popd

function cleanUp {
  local status=$?
  bosh -n delete deployment ${SCALEIO_VATS_DEPLOYMENT_NAME}
  bosh -n delete release ${REXRAY_RELEASE_NAME} || true
  bosh -n delete release ${SCALEIO_SDC_RELEASE_NAME}
  exit $status
}
trap cleanUp EXIT

bosh -n deploy

function ssh_run() {
  sshpass -p ${DEPLOYMENT_PASSWORD} ssh -o "StrictHostKeyChecking no" vcap@${SCALEIO_VSPHERE_VM_IP} \
    "echo ${DEPLOYMENT_PASSWORD} | sudo -S bash -c ' $* '"
}

cat > config.json <<EOF
{
  "volman_driver_path": "/etc/docker/plugins",
  "driver_address": "${VOLUME_DRIVER_ADDRESS}",
  "driver_name": "rexray",
  "create_config": {
    "Name": "${FAKE_VOLUME_NAME}",
    "Opts": {}
  }
}
EOF

sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no config.json vcap@${SCALEIO_VSPHERE_VM_IP}:/home/vcap/

cat > run_test.sh <<EOF
set -x

add-apt-repository -y ppa:ubuntu-lxc/lxd-stable
apt-get -y update
apt-get -y install golang
apt-get -y install git

mkdir -p gocode
export GOPATH=/home/vcap/gocode
export PATH=\$PATH:\$GOPATH/bin

go get --insecure -f -u gopkg.in/yaml.v2
go get --insecure -f -u github.com/onsi/ginkgo/ginkgo
go get --insecure -f -u github.com/onsi/gomega
go get --insecure -f -u github.com/cloudfoundry-incubator/volume_driver_cert
go get --insecure -f -u code.cloudfoundry.org/clock
go get --insecure -f -u code.cloudfoundry.org/cfhttp
go get --insecure -f -u code.cloudfoundry.org/cfhttp/handlers
go get --insecure -f -u github.com/cloudfoundry/gunk/http_wrap
cd \$GOPATH/src/github.com/cloudfoundry-incubator/volume_driver_cert

export FIXTURE_FILENAME=/home/vcap/config.json
ginkgo -r
EOF

chmod +x run_test.sh
sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no run_test.sh vcap@${SCALEIO_VSPHERE_VM_IP}:/home/vcap/

ssh_run ./run_test.sh


ci/templates/run_vats.sh
set -x

add-apt-repository -y ppa:ubuntu-lxc/lxd-stable
apt-get -y update

#install golang
wget https://storage.googleapis.com/golang/go1.7.linux-amd64.tar.gz --no-check-certificate
sudo tar -xvf go1.7.linux-amd64.tar.gz
sudo chmod 777 -R go
sudo mv go /usr/local
export GOROOT=/usr/local/go

apt-get -y install git

mkdir -p gocode
export GOPATH=/home/vcap/gocode
export PATH=$PATH:$GOPATH/bin:$GOROOT/bin

go get --insecure -f -u code.cloudfoundry.org/cfhttp
go get --insecure -f -u code.cloudfoundry.org/cfhttp/handlers
go get --insecure -f -u code.cloudfoundry.org/clock
go get --insecure -f -u code.cloudfoundry.org/lager
go get --insecure -f -u code.cloudfoundry.org/lager/lagertest
go get --insecure -f -u code.cloudfoundry.org/voldriver
go get --insecure -f -u code.cloudfoundry.org/voldriver/driverhttp
go get --insecure -f -u github.com/onsi/ginkgo/ginkgo
go get --insecure -f -u github.com/onsi/gomega
go get --insecure -f -u github.com/tedsuo/ifrit
go get --insecure -f -u github.com/tedsuo/ifrit/ginkgomon
go get --insecure -f -u gopkg.in/yaml.v2

cd $GOPATH/src/code.cloudfoundry.org
git clone https://github.com/EMC-Dojo/volume_driver_cert
cd volume_driver_cert

export FIXTURE_FILENAME=/home/vcap/rexray_config.json
ginkgo -r


ci/tasks/promote-rexray-candidate.sh
#!/usr/bin/env bash

set -e -x

source rexray-bosh-release/ci/tasks/utils.sh

check_param GITHUB_USER
check_param GITHUB_EMAIL
check_param S3_ACCESS_KEY_ID
check_param S3_SECRET_ACCESS_KEY
check_param REXRAY_RELEASE_NAME

# Creates an integer version number from the semantic version format
# May be changed when we decide to fully use semantic versions for releases
integer_version=`cut -d "." -f1 rexray-release-version-semver/number`
echo ${integer_version} > promote/integer_version

cp -r rexray-bosh-release promote/rexray-bosh-release
pushd promote/rexray-bosh-release
  set +x
  echo creating config/private.yml with blobstore secrets
  cat > config/private.yml <<EOF
---
blobstore:
  s3:
    bucket_name: rexray-bosh-release
    access_key_id: ${S3_ACCESS_KEY_ID}
    secret_access_key: ${S3_SECRET_ACCESS_KEY}
EOF
  set -x

  echo "using bosh CLI version..."
  bosh version

  echo "finalizing rexray release..."
  echo '' | bosh create release --force --with-tarball --version ${integer_version} --name ${REXRAY_RELEASE_NAME}
  bosh finalize release dev_releases/${REXRAY_RELEASE_NAME}/*.tgz --version ${integer_version}

  rm config/private.yml

  git diff | cat
  git add .

  git config --global user.email ${GITHUB_EMAIL}
  git config --global user.name ${GITHUB_USER}
  git config --global push.default simple

  git commit -m ":airplane: New final release v ${integer_version} [ci skip]"

popd


ci/tasks/promote-scaleio-sdc-candidate.sh
#!/usr/bin/env bash

set -e -x

source rexray-bosh-release/ci/tasks/utils.sh

check_param GITHUB_USER
check_param GITHUB_EMAIL
check_param S3_ACCESS_KEY_ID
check_param S3_SECRET_ACCESS_KEY
check_param SCALEIO_SDC_RELEASE_NAME

# Creates an integer version number from the semantic version format
# May be changed when we decide to fully use semantic versions for releases
integer_version=`cut -d "." -f1 scaleio-sdc-release-version-semver/number`
echo ${integer_version} > promote/integer_version

cp -r scaleio-sdc-bosh-release promote/scaleio-sdc-bosh-release
pushd promote/scaleio-sdc-bosh-release
  set +x
  echo creating config/private.yml with blobstore secrets
  cat > config/private.yml <<EOF
---
blobstore:
  s3:
    bucket_name: scaleio-sdc-bosh-release
    access_key_id: ${S3_ACCESS_KEY_ID}
    secret_access_key: ${S3_SECRET_ACCESS_KEY}
EOF

  set -x

  echo "using bosh CLI version..."
  bosh version

  echo "finalizing scaleio sdc release..."
  echo '' | bosh create release --force --with-tarball --version ${integer_version} --name ${SCALEIO_SDC_RELEASE_NAME}
  bosh finalize release dev_releases/${SCALEIO_SDC_RELEASE_NAME}/*.tgz --version ${integer_version}

  rm config/private.yml

  git diff | cat
  git add .

  git config --global user.email ${GITHUB_EMAIL}
  git config --global user.name ${GITHUB_USER}
  git config --global push.default simple

  git commit -m ":airplane: New final release v ${integer_version} [ci skip]"

popd


ci/tasks/vats.sh
#!/usr/bin/env bash

set -e -x
source rexray-bosh-release/ci/tasks/utils.sh

check_param BOSH_DIRECTOR_PUBLIC_IP
check_param BOSH_PASSWORD
check_param BOSH_USER
check_param DEPLOYMENT_NAME
check_param DEPLOYMENT_PASSWORD
check_param DEPLOYMENT_PRIVATE_KEY
check_param FAKE_VOLUME_NAME
check_param REXRAY_RELEASE_NAME
check_param SCALEIO_ENDPOINT
check_param SCALEIO_INSECURE
check_param SCALEIO_MDM_IPS
check_param SCALEIO_PASSWORD
check_param SCALEIO_PROTECTION_DOMAIN_ID
check_param SCALEIO_PROTECTION_DOMAIN_NAME
check_param SCALEIO_SDC_RELEASE_NAME
check_param SCALEIO_STORAGE_POOL_ID
check_param SCALEIO_STORAGE_POOL_NAME
check_param SCALEIO_SYSTEM_ID
check_param SCALEIO_USER_ID
check_param SCALEIO_USERNAME
check_param SCALEIO_VERSION
check_param VSPHERE_VM_IP

bosh target ${BOSH_DIRECTOR_PUBLIC_IP}
bosh login ${BOSH_USER} ${BOSH_PASSWORD}

cat > scaleio-acceptance-manifest.yml <<EOF
---
name: scaleio_rexray
director_uuid: cd0eb8bc-831e-447d-99c1-9658c76e7721
stemcells:
- alias: trusty
  os: ubuntu-trusty
  version: '3215'
releases:
- name: rexray-bosh-release
  version: latest
- name: scaleio-sdc-bosh-release
  version: latest
jobs:
- name: scaleio_rexray
  instances: 1
  templates:
  - name: setup_sdc
    release: scaleio-sdc-bosh-release
  - name: rexray_service
    release: rexray-bosh-release
  vm_type: medium
  stemcell: trusty
  azs:
  - z1
  networks:
  - name: private
    static_ips:
    - ${VSPHERE_VM_IP}
  properties:
    network_name: private
properties:
  scaleio:
    mdm:
      ips: ${SCALEIO_MDM_IPS}
  rexray: |
    ---
    rexray:
      modules:
        default-admin:
          desc: The default admin module.
          disabled: false
          host: tcp://127.0.0.1:7979
          type: admin
        default-docker:
          desc: The default docker module.
          disabled: false
          host: tcp://127.0.0.1:9000
          spec: /etc/docker/plugins/rexray.spec
          type: docker
      libstorage:
        embedded: true
        driver: scaleio
        service: scaleio
    scaleio:
      endpoint: ${SCALEIO_ENDPOINT}
      insecure: ${SCALEIO_INSECURE}
      password: ${SCALEIO_PASSWORD}
      protectionDomainName: default
      storagePoolName: ${SCALEIO_STORAGE_POOL_NAME}
      systemID: ${SCALEIO_SYSTEM_ID}
      thinOrThick: ThinProvisioned
      useCerts: false
      userID: ${SCALEIO_USER_ID}
      userName: ${SCALEIO_USERNAME}
      version: ${SCALEIO_VERSION}
    linux:
      volume:
        fileMode: 0777
update:
  canaries: 1
  max_in_flight: 3
  canary_watch_time: 30000-600000
  update_watch_time: 5000-600000

EOF
cat scaleio-acceptance-manifest.yml
bosh deployment scaleio-acceptance-manifest.yml

apt-get -y update && apt-get -y install sshpass

pushd scaleio-sdc-bosh-release
  bosh -n create release --force --name ${SCALEIO_SDC_RELEASE_NAME}
  bosh -n upload release
popd

pushd rexray-bosh-release
  bosh -n create release --force --name ${REXRAY_RELEASE_NAME}
  bosh -n upload release
popd

function cleanUp {
  local status=$?
  bosh -n delete deployment ${DEPLOYMENT_NAME}
  bosh -n delete release ${REXRAY_RELEASE_NAME}
  bosh -n delete release ${SCALEIO_SDC_RELEASE_NAME}
  exit $status
}
trap cleanUp EXIT

bosh -n deploy

function ssh_run() {
  sshpass -p ${DEPLOYMENT_PASSWORD} ssh -o "StrictHostKeyChecking no" vcap@${VSPHERE_VM_IP} \
    "echo ${DEPLOYMENT_PASSWORD} | sudo -S bash -c ' $* '"
}

cat > config.json <<EOF
{
  "volman_driver_path": "/etc/docker/plugins",
  "driver_name": "rexray",
  "create_config": {
    "Name": "${FAKE_VOLUME_NAME}",
    "Opts": {}
  }
}
EOF

sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no config.json vcap@${VSPHERE_VM_IP}:/home/vcap/

cat > run_test.sh <<EOF
set -x
wget --no-check-certificate https://storage.googleapis.com/golang/go1.6.2.linux-amd64.tar.gz
tar -zxf go1.6.2.linux-amd64.tar.gz -C ./
export GOROOT=/home/vcap/go
export GOOS=linux
export GOARCH=amd64
export PATH=\$PATH:\$GOROOT/bin

mkdir -p gocode
export GOPATH=/home/vcap/gocode
export PATH=\$PATH:\$GOPATH/bin

/var/vcap/packages/rexray/rexray volume create --volumename ${FAKE_VOLUME_NAME} --size 8 || true

apt-get -y update && apt-get -y install git && apt-get -y install jq

mkdir -p \$GOPATH/src/gopkg.in/
cd \$GOPATH/src/gopkg.in/
git clone https://github.com/go-yaml/yaml.git yaml.v2
go install gopkg.in/yaml.v2

mkdir -p \$GOPATH/src/github.com/onsi
cd \$GOPATH/src/github.com/onsi

git clone https://github.com/onsi/ginkgo.git
git clone https://github.com/onsi/gomega.git
go install github.com/onsi/ginkgo/ginkgo
go install github.com/onsi/gomega

go get --insecure -f -u github.com/cloudfoundry-incubator/volume_driver_cert
go get --insecure -f -u code.cloudfoundry.org/clock
cd \$GOPATH/src/github.com/cloudfoundry-incubator/volume_driver_cert

printf "http://127.0.0.1:9000" > /etc/docker/plugins/rexray.spec

export FIXTURE_FILENAME=/home/vcap/config.json
ginkgo -r

/var/vcap/packages/rexray/rexray volume unmount --volumename ${FAKE_VOLUME_NAME} || true
/var/vcap/packages/rexray/rexray volume remove --volumeid $(/var/vcap/packages/rexray/rexray volume get --volumename ${FAKE_VOLUME_NAME} -f json | jq .id -r)
EOF

chmod +x run_test.sh
sshpass -p ${DEPLOYMENT_PASSWORD} scp -o StrictHostKeyChecking=no run_test.sh vcap@${VSPHERE_VM_IP}:/home/vcap/

ssh_run ./run_test.sh
